Sources/System/MachPort.swift

```diff
@@ -0,0 +1,318 @@
+    }
+
+    /// Allocate a new Mach port with a receive right, creating a
+    /// Mach.Port<Mach.ReceiveRight> to manage it.
+    ///
+    /// This initializer will abort if the right could not be created.
+    /// Callers may assert that a valid right is always returned.
+    init() {
+        var storage:mach_port_name_t = 0
+        withUnsafeMutablePointer(to:&storage) { storage in
+            let kr = mach_port_allocate(mach_task_self_, MACH_PORT_RIGHT_RECEIVE, storage)
+            assert(kr == KERN_SUCCESS)
+        }
+
+        // name-only init will guard ReceiveRights
+        self.init(name:storage)
+    }
+
+
+    /// Transfer ownership of the underlying port right to the caller.
+    ///
+    /// Returns a tuple containing the Mach port name representing the right,
+    /// and the context value used to guard the right.
+    ///
+    /// This operation liberates the right from management by the Mach.Port,
+    /// and the underlying right will no longer be automatically deallocated.
+    ///
+    /// After this function completes, the Mach.Port is destroyed and no longer
+    /// usable.
+    __consuming func relinquish() -> (mach_port_name_t, mach_port_context_t) {
+        return (name, context)
+    }
+
+    /// Remove guard and transfer ownership of the underlying port right to
+    /// the caller.
+    ///
+    /// Returns the Mach port name representing the right.
+    ///
+    /// This operation liberates the right from management by the Mach.Port,
+    /// and the underlying right will no longer be automatically deallocated.
+    ///
+    /// After this function completes, the Mach.Port is destroyed and no longer
+    /// usable.
+    ///
+    /// This function makes a syscall to remove the guard from
+    /// Mach.ReceiveRights. Use relinquish() to avoid the syscall and extract
+    /// the context value along with the port name.
+    __consuming func unguardAndRelinquish() -> mach_port_name_t {
+        let kr = mach_port_unguard(mach_task_self_, name, context);
+        assert(kr == KERN_SUCCESS)
+        return name
+    }
+
+    /// Borrow access to the port name in a block that can perform
+    /// non-consuming operations.
+    ///
+    /// Take care when using this function; many operations consume rights.
+    ///
+    /// If the right is consumed, behavior is undefined.
+    ///
+    /// The body block may optionally return something, which will then be
+    /// returned to the caller of withBorrowedName.
+    func withBorrowedName<ReturnType>(body:(mach_port_name_t, mach_port_context_t) -> ReturnType) -> ReturnType {
+        body(name, context)
+    }
+
+    /// Create a send-once right for a given receive right.
+    ///
+    /// This does not affect the makeSendCount of the receive right.
+    ///
+    /// This function will abort if the right could not be created.
+    /// Callers may assert that a valid right is always returned.
+    func makeSendOnceRight() -> Mach.Port<Mach.SendOnceRight> {
```

Comment:
Yes, but only from receive rights, because this interface makes running out of ipc space or uref overflow fatal, and invalid name/right mean that either the Mach.Port has not successfully managed the right or the contract with Mach.Port was violated (the right being destroyed out from under it).